gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/actors/CheckPermissionsActor.java

```diff
@@ -0,0 +1,417 @@
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.ditto.gateway.service.endpoints.actors;
+
+import java.time.Duration;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+import org.apache.pekko.actor.AbstractActor;
+import org.apache.pekko.actor.ActorRef;
+import org.apache.pekko.actor.Props;
+import org.apache.pekko.japi.pf.ReceiveBuilder;
+import org.apache.pekko.pattern.Patterns;
+import org.eclipse.ditto.base.model.exceptions.DittoInternalErrorException;
+import org.eclipse.ditto.base.model.exceptions.DittoRuntimeException;
+import org.eclipse.ditto.base.model.headers.DittoHeaders;
+import org.eclipse.ditto.gateway.service.endpoints.routes.checkpermissions.CheckPermissions;
+import org.eclipse.ditto.gateway.service.endpoints.routes.checkpermissions.CheckPermissionsResponse;
+import org.eclipse.ditto.gateway.service.endpoints.routes.checkpermissions.ImmutablePermissionCheck;
+import org.eclipse.ditto.gateway.service.endpoints.routes.checkpermissions.PermissionCheckWrapper;
+import org.eclipse.ditto.internal.utils.pekko.logging.DittoLogger;
+import org.eclipse.ditto.internal.utils.pekko.logging.DittoLoggerFactory;
+import org.eclipse.ditto.json.JsonFieldSelector;
+import org.eclipse.ditto.policies.model.PolicyId;
+import org.eclipse.ditto.policies.model.ResourceKey;
+import org.eclipse.ditto.policies.model.ResourcePermissionFactory;
+import org.eclipse.ditto.policies.model.ResourcePermissions;
+import org.eclipse.ditto.policies.api.commands.sudo.CheckPolicyPermissions;
+import org.eclipse.ditto.policies.api.commands.sudo.CheckPolicyPermissionsResponse;
+import org.eclipse.ditto.things.api.commands.sudo.SudoRetrieveThing;
+import org.eclipse.ditto.things.api.commands.sudo.SudoRetrieveThingResponse;
+import org.eclipse.ditto.things.model.Thing;
+import org.eclipse.ditto.things.model.ThingId;
+
+/**
+ * Actor that handles permission checks for resources and entities.
+ * <p>
+ * This actor receives {@link CheckPermissions} messages, processes permission checks by
+ * aggregating results based on policies and entities, and communicates with the Ditto Edge
+ * and Things API to retrieve policy data and permissions.
+ * It responds to the sender with a {@link CheckPermissionsResponse} after evaluating the
+ * permission checks.
+ * <p>
+ * The actor forwards requests to the edge command forwarder to retrieve policies and permission
+ * results, handling various entity and policy resources.
+ *
+ * @since 3.7.0
+ */
+final class CheckPermissionsActor extends AbstractActor {
```

Comment:
The actor is not shutting down itself.

We must add that, as one actor is created for each request .. this is a potential memory and resource leak, if we don't shut it down again after handling the request.

I will point out the place where to do best..